src/Umbraco.Infrastructure/Services/MemberEditingService.cs

```diff
@@ -246,16 +249,28 @@ private async Task<MemberEditingOperationStatus> ValidateMemberDataAsync(MemberE
         return MemberEditingOperationStatus.Success;
     }
 
-    private async Task<bool> UpdateRoles(IEnumerable<string>? roles, MemberIdentityUser identityMember)
+    private async Task<bool> UpdateRoles(IEnumerable<Guid>? roles, MemberIdentityUser identityMember)
     {
+        // We have to convert the GUIDS to names here, as roles on a member are stored by name, not key.
+        var memberGroups = new List<IMemberGroup>();
+        foreach (Guid key in roles ?? Enumerable.Empty<Guid>())
+        {
+
+            IMemberGroup? group = await _memberGroupService.GetAsync(key);
+            if (group is not null)
+            {
+                memberGroups.Add(group);
+            }
+        }
+
         // We're gonna look up the current roles now because the below code can cause
         // events to be raised and developers could be manually adding roles to members in
         // their handlers. If we don't look this up now there's a chance we'll just end up
         // removing the roles they've assigned.
         IEnumerable<string> currentRoles = (await _memberManager.GetRolesAsync(identityMember)).ToList();
 
         // find the ones to remove and remove them
-        var rolesToRemove = currentRoles.Except(roles ?? Enumerable.Empty<string>()).ToArray();
+        var rolesToRemove = currentRoles.Except(memberGroups.Select(x => x.Name)).WhereNotNull().ToArray();
```

Comment:
Instead of calling `memberGroups.Select(x => x.Name)` and then `memberGroups.Select(x => x.Name).WhereNotNull()` I think it would be to just call `memberGroups.Select(x => x.Name).WhereNotNull()` once and store it in a variable that can be reused, that way we don't have to do additional enumerations.